arch-arm: add Sve mla and mls indexed

Nitesh8998 opened the pull request:
This contains the implementation of mla and MLS index version instructions from ARM SVE2 ISA specification.

Nitesh8998 committed:
arch-arm: This commit adds the mla/s indexed versions

This includes the isa and instruction implementations
of mla and mls indexed versions from ARM SVE2 ISA spec.

Change-Id: I4fbd0382f23d8611e46411f74dc991f5a211a313

powerjg commented:
@giactra are you going to be able to review this before the staging branch is created on Friday?

andysan commented:
> @giactra are you going to be able to review this before the staging branch is created on Friday?

We'll need to run some arch compliance tests on this. Unfortunately, I don't think we'll be able to do that before the weekend.

powerjg commented:
Moving this to *not* in 23.1, then.

powerjg commented:
@andysan now that we have more testing bandwidth, can we make these compliance tests public? It would improve the workflow for people if they could see test results and we could run them.

andysan commented:
> @andysan now that we have more testing bandwidth, can we make these compliance tests public? It would improve the workflow for people if they could see test results and we could run them.

Sorry, that's unfortunately not possible. They are produced by our internal test teams and not suitable for public consumption.

giactra commented:
Just came back from a long vacation; will find some time today to review the PR

giactra commented on src/arch/arm/insts/sve.hh:
```diff
@@ -498,6 +498,27 @@ class SveTerPredOp : public ArmStaticInst
+  protected:
+    RegIndex dest, op1, op2;
+    uint16_t imm;
+    uint8_t esize;
```
Other than esize (which is not necessary), this is the same as SveBinIdxUnpredOp. I would suggest implementing MLA/S (idx) on top of SveBinIdxUnpredOp

giactra commented on src/arch/arm/isa/templates/sve.isa:
```diff
@@ -568,6 +606,8 @@ class %(class_name)s : public %(base_class)s
 };
 }};
 
+
+
```
Please remove these newlines

giactra commented on src/arch/arm/isa/insts/sve.isa:
```diff
@@ -3094,6 +3122,9 @@ let {{
             substDict = {'targs': type, 'class_name': 'Sve' + Name}
             exec_output += SveOpExecDeclare.subst(substDict)
 
+
+
+
```
Please remove these newlines

giactra commented on src/arch/arm/isa/formats/sve_2nd_level.isa:
```diff
@@ -245,6 +245,59 @@ namespace Aarch64
+        RegIndex zn = (RegIndex) (uint8_t) bits(machInst, 9, 5);
+        RegIndex zm ;
+
+        //= (RegIndex) (uint8_t) bits(machInst, 19, 16);
+
```
You could remove this

giactra commented on src/arch/arm/isa/formats/sve_2nd_level.isa:
```diff
@@ -245,6 +245,59 @@ namespace Aarch64
+       switch(size) {
+            case 0b00:
+            case 0b01:
+                zm = (RegIndex)(uint8_t)bits(machInst, 18, 16);
+                imm = (uint16_t)(bits(machInst, 22) << 2)
```
I undertand zm and imm depend on the element type; may I suggest you to define them within the switch case?
Something like:

       switch(size) {
            case 0b00:
            case 0b01: {
                RegIndex zm = (RegIndex)(uint8_t)bits(machInst, 18, 16);
                RegIndex imm = (uint16_t)(bits(machInst, 22) << 2)

                [...]
            }



I would like to avoid having some code using uninitialized variables by accident

giactra commented on src/arch/arm/isa/templates/sve.isa:
```diff
@@ -515,6 +515,44 @@ class %(class_name)s : public %(base_class)s
 }};
 
+
+
+def template SveTerIndexedOpDeclare {{
+  /*
+  For mla indexed version as it is not included in gem5 right now.
+  Using ternary ops but all ops are vector regs.
+  index is the imm here. (name can be changed)
+*/
+template <class _Element>
+class %(class_name)s : public %(base_class)s
+{
+  //static_assert(sizeof(_SElementA) == sizeof(_SElementB),
+  //             "Source elements must have the same size.");
+
+  private:
+    %(reg_idx_arr_decl)s;
+
+  protected:
+    typedef _Element Element;
+    typedef _Element TPElem;
+
+  public:
+    // Constructor
+    %(class_name)s(ExtMachInst machInst, RegIndex _dest,
+                   RegIndex _op1, RegIndex _op2, uint16_t _imm) :
+        %(base_class)s("%(mnemonic)s", machInst, %(op_class)s,
+                       _dest, _op1, _op2,_imm)
+    {
+        %(set_reg_idx_arr)s;
+        %(constructor)s;
+        esize = sizeof(Element);
+    }
+
+    Fault execute(ExecContext *, trace::InstRecord *) const override;
+};
+}};
+
+
```
If we use sveBinIdxInst this won't be needed as the python generator will automatically pick SveBinIdxUnpredOpDeclare which does its job

giactra commented on src/arch/arm/isa/insts/sve.isa:
```diff
@@ -2096,6 +2096,34 @@ let {{
                          'class_name' : 'Sve' + Name}
             exec_output += SveOpExecDeclare.subst(substDict)
 
+    # Generates definitions for ternary SVE instructions (indexed)
+    def sveTerInstIndexed(name, Name, opClass, types, op, decoder='Generic'):
+        global header_output, exec_output, decoders
+        code = sveEnabledCheckCode + '''
+        unsigned eCount = ArmStaticInst::getCurSveVecLen<Element>(
+                xc->tcBase());
+        for (unsigned i = 0; i < eCount; i++) {
+            int segbase = i - i % (128 / sizeof(Element));
+            int s = segbase + imm;
+            const Element& srcElem1 = AA64FpOp1_x[i];
+            const Element& srcElem2 = AA64FpOp2_x[s];
+            Element destElem = AA64FpDestMerge_x[i];
+            '''
+
+        code +=  f"{op} \n"
+
+
+        code += ''' AA64FpDest_x[i] = destElem;
+        }'''
+        iop = ArmInstObjParams(name, 'Sve2' + Name+ 'i', 'SveTerIndexedOp',
+                               {'code': code, 'op_class': opClass}, [])
+        header_output += SveTerIndexedOpDeclare.subst(iop)
+        exec_output += SveOpExecute.subst(iop)
+        for type in types:
+            substDict = {'targs' : type,
+                         'class_name' : 'Sve2' + Name + "i"}
+            exec_output += SveOpExecDeclare.subst(substDict)
+
```
This is effectively the same as sveBinIdxInst.. I believe we should reuse existing code

Nitesh8998 replied:
Hi,

Thank you for your suggestions, while I make the changes I have a few observations I would like to bring to you.
the sveBinIdxInst function, here would might not work as it does the following 

```
Element destElem = 0;
```
Instead of 
```
Element destElem = AA64FpDestMerge_x[i];
```

which is needed for mla/s indexed. However, as you mentioned the template library does have some repeated code and I see a  ```sveTerIdxInst``` function that existed before and solves the purpose I shall use that.


Also, I have another observation and would like to know more about why the implementation is done this way. For both ```sveTerIdxInst``` and ```sveBinIdxInst``` there is a temporary  ```VegRegContainer``` created as ```auxDest``` ( seems like auxiliary destination), which then is written into ``` AA64FpDest_x[i] ``` using a separate for loop as below:

```
    Element destElem = AA64FpDestMerge_x[i];
        '''
        code += '''
            %(op)s
            auxDest[i] = destElem;
        }

        for (unsigned i = 0; i < eCount; i++) {
            AA64FpDest_x[i] = auxDest[i];
        }''' % {'op': op}
```

Is there a reason why this is done instead of using ```AA64FpDest_x[i]``` in the previous for-loop (like other instructions that use the ```sveTerInst``` function?

Let me know.

Regards

giactra replied:
> Hi,
> 
> Thank you for your suggestions, while I make the changes I have a few observations I would like to bring to you. the sveBinIdxInst function, here would might not work as it does the following
> 
> ```
> Element destElem = 0;
> ```
> 
> Instead of
> 
> ```
> Element destElem = AA64FpDestMerge_x[i];
> ```
> 

Oh right I missed that!

> which is needed for mla/s indexed. However, as you mentioned the template library does have some repeated code and I see a `sveTerIdxInst` function that existed before and solves the purpose I shall use that.
> 

That's great!

> Also, I have another observation and would like to know more about why the implementation is done this way. For both `sveTerIdxInst` and `sveBinIdxInst` there is a temporary `VegRegContainer` created as `auxDest` ( seems like auxiliary destination), which then is written into `AA64FpDest_x[i]` using a separate for loop as below:
> 
> ```
>     Element destElem = AA64FpDestMerge_x[i];
>         '''
>         code += '''
>             %(op)s
>             auxDest[i] = destElem;
>         }
> 
>         for (unsigned i = 0; i < eCount; i++) {
>             AA64FpDest_x[i] = auxDest[i];
>         }''' % {'op': op}
> ```
> 
> Is there a reason why this is done instead of using `AA64FpDest_x[i]` in the previous for-loop (like other instructions that use the `sveTerInst` function?
> 
> Let me know.
> 
> Regards

This could potentially be an unnecessary copy. I have two possible explanations:

1) To handle the case where the source = destination register, meaning

`AA64FpDest_x[i] = destElem`

might change the source register value while still iterating. This is not the case anyway, as while writing to the destination register changes the physical register, the source operands are copied anyway into temporaries tmp_s* with getRegOperand

2) To handle FP exceptions thrown while iterating over the vectors. In case of an exception we don't want to have an inconsistent register state.. FP exceptions are not implemented in Arm anyway so again, there seem to be non need for that.

I will run some tests and see if we can safely remove that auxiliary vector.. Anyway it should be indipendent from this PR. Thanks for flagging it though!

Nitesh8998 replied:
Thank you for clearing the doubts, do let me know if I can do anything regarding this. I would like to be involved with it.

giactra reviewed (changes requested):
First of all thanks for your contribution and apologies if it took me some time to review it.

I believe it looks good from a functional perspective but there is a lot of duplication with existing code. This is not your fault; the template library has grown chaotically and the chance of duplicating classes and functions is extremely high. Please check my comments to see what can be rewritten with existing logic

Nitesh8998 committed:
arch-arm: This commit uses existing template code for mla/s index

This includes mla/s index version  implementation using the existing template code
to avoid code repeatition.

Change-Id: If1de84e01dec638e206c979ca832308ebc904212

giactra commented on src/arch/arm/isa/templates/sve.isa:
```diff
@@ -515,6 +515,10 @@ class %(class_name)s : public %(base_class)s
 };
 }};
 
+
+
+
+
```
Please remove newlines

giactra commented on src/arch/arm/isa/templates/sve.isa:
```diff
@@ -568,6 +572,8 @@ class %(class_name)s : public %(base_class)s
 };
 }};
 
+
+
```
Ditto

giactra commented on src/arch/arm/isa/insts/sve.isa:
```diff
@@ -3094,6 +3094,9 @@ let {{
             substDict = {'targs': type, 'class_name': 'Sve' + Name}
             exec_output += SveOpExecDeclare.subst(substDict)
 
+
+
+
```
Please remove newlines

giactra reviewed (changes requested):
Thanks Nitesh, it looks good to me, some minor coding style issues to fix (basically removing newlines addition) and then we are free to merge this

Nitesh8998 committed:
arch-arm: This commit cleans .isa files

This commit cleans extra new lines from .isa files from this branch

Change-Id: I4087ed230aa041747038b49360c2aba3f82c0790
